<h1 style='color: blue; font-size: 34px; font-weight: bold;'> Projeto Proposto 
</h1>

<p style='font-size: 18px; line-height: 2; margin: 0px 0px; text-align: justify; text-indent: 0px;'>    
<i> Este projeto baseia-se no Estudo das Técnicas de Deep Learning - Redes Neurais Artificiais. </i>       
</p>  

> CONSULTAR :

https://didatica.tech/introducao-a-redes-neurais-e-deep-learning/

https://www.youtube.com/watch?v=GByl_PEo94M&list=PLFE-LjWAAP9R4G0WOXWuha4P5cCvw7hGB&index=16&ab_channel=DataICMC

https://medium.com/ensina-ai/redes-neurais-roots-1-introdu%C3%A7%C3%A3o-ffdd6f8b9f01

https://medium.com/ensina-ai/redes-neurais-roots-2-treinamento-3161a439c4f3

https://medium.com/ensina-ai/redes-neurais-roots-3-show-me-the-code-e56359310083

https://www.youtube.com/watch?v=fEukSrpDPH0&list=PLtQM10PgmGogjn0cikgWi8wpQUnV6ERkY&index=9&ab_channel=DiogoCortiz

https://www.youtube.com/watch?v=0eQ6-6A5W6I&list=PLtQM10PgmGogjn0cikgWi8wpQUnV6ERkY&index=10&ab_channel=DiogoCortiz

https://www.youtube.com/watch?v=WPxV5WXP3OM&list=PL-t7zzWJWPtygNTsgC_M8c9a-p5biCjho&ab_channel=UniversoDiscreto

https://www.youtube.com/watch?v=CqOfi41LfDw&list=PLblh5JKOoLUIxGDQs4LFFD--41Vzf-ME1&ab_channel=StatQuestwithJoshStarmer

# <font color='red' style='font-size: 40px;'> Library  </font>

In [1]:
import pandas as pd 
import numpy as np
import scipy as sp 
import matplotlib.pyplot as plt
import seaborn as sns 
import warnings



%matplotlib inline
sns.set()
plt.rcParams['font.family'] = 'Arial'
plt.rcParams['font.size'] = '14'
plt.rcParams['figure.figsize'] = [10, 5]
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 100)
pd.set_option('display.float_format', lambda x: '%.2f' % x) # Tira os números do formato de Notação Científica
np.set_printoptions(suppress=True) # Tira os números do formato de Notação Científica em Numpy Arrays
warnings.filterwarnings('ignore') # Remove Warnings
warnings.simplefilter(action='ignore', category=FutureWarning) # Retira Future Warnings

# <font color='red' style='font-size: 40px;'> 1.0) Introdução as Redes Neurais </font>

## 1.1) Introdução às Redes Neurais

<img src="./img01.png" width="70%">
<img src="./img02.png" width="70%">
<img src="./img03.png" width="70%">
<img src="./img04.png" width="70%">
<img src="./img05.png" width="70%">

# <font color='red' style='font-size: 40px;'> 2.0) Redes Neurais Artificiais - Conceitos Teóricos </font>

## 2.1) Neurônio Biológico, Neurônio Artificial e Redes Neurais

<img src="./img06.png" width="70%">
<img src="./img07.png" width="70%">
<img src="./img08.png" width="70%">
<img src="./img09.png" width="70%">
<img src="./img10.png" width="70%">
<img src="./img11.png" width="70%">
<img src="./img12.png" width="70%">
<img src="./img13.png" width="70%">


## 2.2) Principais Tipos de Redes Neurais 

<img src="./img14.png" width="70%">

## 2.3) Perceptron

<img src="./img15.png" width="70%">
<img src="./img16.png" width="70%">
<img src="./img17.png" width="70%">
<img src="./img18.png" width="70%">




## 2.4) Função de Ativação 

#### Sendo cada Neurônio uma "combinação": X1 (Neurônio Gênero Masculino), X2 (Neurônio Gênero Feminino), X3 (Neurônio Estado Civil Casado), X4 (Neurônio Estado Civil Viúva), X6 (Neurônio Estado Civil Solteiro)...Xn (Neurônio i,j)
<img src="./img19.png" width="70%">
<img src="./img20.png" width="70%">
<img src="./img21.png" width="70%">
<img src="./img22.png" width="70%">
<img src="./img23.png" width="70%">
<img src="./img24.png" width="70%">
<img src="./img25.png" width="70%">
<img src="./img26.png" width="70%">
<img src="./img27.png" width="70%">




## 2.5) Aprendizagem nas Redes Neurais Artificiais 

<img src="./img27.png" width="70%">
<img src="./img28.png" width="70%">
<img src="./img29.png" width="70%">
<img src="./img30.png" width="70%">
<img src="./img31.png" width="70%">
<img src="./img32.png" width="70%">
<img src="./img33.png" width="70%">
<img src="./img34.png" width="70%">




## 2.6) Aprendizagem com Gradiente Descendente 

<img src="./img36.png" width="70%">
<img src="./img35.png" width="70%">
<img src="./img37.png" width="70%">
<img src="./img38.png" width="70%">
<img src="./img39.png" width="70%">




## 2.7) Cálculo dos Erros

<img src="./img40.png" width="70%">
<img src="./img41.png" width="70%">
<img src="./img42.png" width="70%">
<img src="./img43.png" width="70%">


## 2.8) Topologia das Redes Neurais  

<img src="./img44.png" width="70%">
<img src="./img45.png" width="70%">
<img src="./img46.png" width="70%">
<img src="./img47.png" width="70%">
<img src="./img48.png" width="70%">

## 2.9) Algoritmo Backpropagation

<img src="./img49.png" width="70%">
<img src="./img50.png" width="70%">


## 2.10) Definição dos Parâmetros 

<img src="./img50.png" width="70%">
<img src="./img51.png" width="70%">
<img src="./img52.png" width="70%">
<img src="./img53.png" width="70%">
<img src="./img54.png" width="70%">
<img src="./img55.png" width="70%">
<img src="./img56.png" width="70%">


# <font color='red' style='font-size: 40px;'> 3.0) Redes Neurais Artificiais - Classificação </font>



## **Análise Socioeconômica do Estado de Minas Gerais**

Este projeto tem por objetivo utilizar Algoritmos de Redes Neurais Artificiais para realizar uma Análise Socioeconômica do Estado de Minas Gerais utilizando as respostas dos questionários dos inscritos no Enem de 2019.

Para este caso, o Dataset já passou por um Preparação e Tratamento anteriores.  

In [20]:
import pandas as pd 
import numpy as np
import scipy as sp 
import matplotlib.pyplot as plt
import seaborn as sns 
import warnings
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler
from sklearn.model_selection import train_test_split, StratifiedKFold, cross_val_score
from sklearn.compose import ColumnTransformer
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

%matplotlib inline
sns.set()
plt.rcParams['font.family'] = 'Arial'
plt.rcParams['font.size'] = '14'
plt.rcParams['figure.figsize'] = [10, 5]
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 100)
pd.set_option('display.float_format', lambda x: '%.2f' % x) # Tira os números do formato de Notação Científica
np.set_printoptions(suppress=True) # Tira os números do formato de Notação Científica em Numpy Arrays
warnings.filterwarnings('ignore') # Remove Warnings
warnings.simplefilter(action='ignore', category=FutureWarning) # Retira Future Warnings

In [3]:
dados_tratados = pd.read_csv('dados_tratados.csv', sep = ',', encoding = 'iso-8859-1')
print(dados_tratados.shape)
print(dados_tratados.dtypes)
dados_tratados.head()

(538896, 18)
Q005     int64
Q006    object
Q007    object
Q008    object
Q009    object
Q010    object
Q011    object
Q012    object
Q014    object
Q015    object
Q016    object
Q017    object
Q019    object
Q021    object
Q022    object
Q023    object
Q024    object
Q025    object
dtype: object


Q005 Q006 Q007 Q008 Q009 Q010 Q011 Q012 Q014 Q015 Q016 Q017 Q019 Q021 Q022  \
0     3    I    A    C    D    B    A    B    B    A    B    A    D    A    D   
1     5    D    A    B    C    A    B    B    A    A    A    A    B    A    C   
2     3    F    A    B    D    A    A    B    B    B    A    A    C    A    C   
3     3    P    B    D    D    C    A    B    B    A    B    A    B    A    D   
4     3    D    A    B    D    B    B    B    A    A    A    A    B    A    C   

  Q023 Q024 Q025  
0    B    B    B  
1    A    A    A  
2    A    A    B  
3    A    D    B  
4    A    B    A

In [4]:
x = dados_tratados.iloc[:, 0:17].values
y = dados_tratados.iloc[:, 17].values

transformacao_x = ColumnTransformer([('transformacao', OneHotEncoder(), [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16])], remainder = 'passthrough')
x = transformacao_x.fit_transform(x).toarray()

transformacao_y = LabelEncoder()
y = transformacao_y.fit_transform(y)

print(f'Variáveis Preditoras: {x.shape}')
print(f'Variável Target: {y.shape}')

Variáveis Preditoras: (538896, 105)
Variável Target: (538896,)


In [9]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.15, random_state = 42)
print(x_train.shape)
print(y_train.shape)
print(x_test.shape)
print(y_test.shape)

(458061, 105)
(458061,)
(80835, 105)
(80835,)


#### Parâmetros MLPClassifier   

- hidden_layer_sizes (camadas escondidas): default (100,)   

  Quant.= (Ne+Ns)/2 = (105+1)/2 = 53 neurônios   
  Quant.=2/3.(Ne) + Ns = 2/3.105+1 = 71 neurônios
- activation: Função de ativação na camada oculta default='relu' 
- solver: algoritmo matemático. Default='adam' (datasets grandes = acima de 1000 amostras). lbfgs é para datasets pequenos. sgd é com a descida do gradiente estocástico (recomendado testar).
- alpha:  parâmetro para o termo de regularização de ajuste de pesos. Aumento de alpha estimula pesos menores e diminuição de alpha estimula pesos maiores. Default=0.0001.
- batch_size: tamanho dos mini lotes. default=min(200, n_samples). Não usar com o solver lbfgs.
- learning_rate: taxa de aprendizagem. default='constant'. Três tipos:   
  1- 'constant':uma taxa de aprendizado constante dada pela taxa de aprendizagem inicial.   
  2- 'invscaling': diminui gradualmente por: taxa efetiva = taxa inicial / t^power_t  
  3- 'adaptive': a taxa é dividida por 5 cada vez que em duas épocas consecutivas não diminuir o erro.
- learning_rate_init: taxa de aprendizagem inicial. Default=0.001
- power_t: default=0.5. usado apenas para solver 'sgd'.
- max_iter int: Número máximo de iterações. default = 200.  ('sgd', 'adam').
- max_fun: Número máximo de chamadas de função de perda. Para 'lbfgs'. Default: 15000
- shuffle: default = True
Usado apenas quando solver = 'sgd' ou 'adam'.
- random_state: default = None
- tol:Tolerância para a otimização.Default=0.0001
- momentum: otimização do algoritmo 'sgd'. Default: 0.9.
- n_iter_no_change: Número máximo de épocas que não atinge a tolerância de melhoria. default = 10. Apenas para solver = 'sgd' ou 'adam'
- verbose : Mostra o progresso. default=False.

In [7]:
MLPClassifier?

Init signature:
MLPClassifier(
    hidden_layer_sizes=(100,),
    activation='relu',
    *,
    solver='adam',
    alpha=0.0001,
    batch_size='auto',
    learning_rate='constant',
    learning_rate_init=0.001,
    power_t=0.5,
    max_iter=200,
    shuffle=True,
    random_state=None,
    tol=0.0001,
    verbose=False,
    warm_start=False,
    momentum=0.9,
    nesterovs_momentum=True,
    early_stopping=False,
    validation_fraction=0.1,
    beta_1=0.9,
    beta_2=0.999,
    epsilon=1e-08,
    n_iter_no_change=10,
    max_fun=15000,
)
Docstring:     
Multi-layer Perceptron classifier.

This model optimizes the log-loss function using LBFGS or stochastic
gradient descent.

.. versionadded:: 0.18

Parameters
----------
hidden_layer_sizes : tuple, length = n_layers - 2, default=(100,)
    The ith element represents the number of neurons in the ith
    hidden layer.

activation : {'identity', 'logistic', 'tanh', 'relu'}, default='relu'
    Activation function for the hidden layer.

  

In [10]:
mlp_classifier = MLPClassifier(hidden_layer_sizes = (53, 53), activation = 'relu', solver = 'adam', alpha = 0.0001, batch_size='auto', 
learning_rate='constant', learning_rate_init=0.001, max_iter = 500, tol = 0.01, random_state = 42, verbose = True)
mlp_classifier.fit(x_train, y_train) 

Iteration 1, loss = 0.30293168
Iteration 2, loss = 0.29870722
Iteration 3, loss = 0.29813710
Iteration 4, loss = 0.29770843
Iteration 5, loss = 0.29736986
Iteration 6, loss = 0.29708370
Iteration 7, loss = 0.29678903
Iteration 8, loss = 0.29655239
Iteration 9, loss = 0.29625942
Iteration 10, loss = 0.29615045
Iteration 11, loss = 0.29588726
Iteration 12, loss = 0.29565604
Training loss did not improve more than tol=0.010000 for 10 consecutive epochs. Stopping.


MLPClassifier(hidden_layer_sizes=(53, 53), max_iter=500, random_state=42,
              tol=0.01, verbose=True)

In [17]:
predict = mlp_classifier.predict(x_test)
print(f'Função de Ativação na Camada de Saída: {mlp_classifier.out_activation_}')
acuracia = accuracy_score(y_test, predict)
matriz_de_confusao = confusion_matrix(y_test, predict)

print(f'Acurácia: {acuracia}')
print(matriz_de_confusao)


Função de Ativação na Camada de Saída: logistic
Acurácia: 0.8698707243149626
[[ 5270  7385]
 [ 3134 65046]]


In [18]:
mlp_classifier = MLPClassifier(hidden_layer_sizes = (53, 53), activation = 'relu', solver = 'sgd', alpha = 0.0001, batch_size='auto', 
learning_rate='constant', learning_rate_init=0.001, max_iter = 500, tol = 0.01, random_state = 42, verbose = True)
mlp_classifier.fit(x_train, y_train) 

Iteration 1, loss = 0.33774431
Iteration 2, loss = 0.30534264
Iteration 3, loss = 0.30264663
Iteration 4, loss = 0.30142329
Iteration 5, loss = 0.30067201
Iteration 6, loss = 0.30021143
Iteration 7, loss = 0.29987414
Iteration 8, loss = 0.29963062
Iteration 9, loss = 0.29944518
Iteration 10, loss = 0.29929903
Iteration 11, loss = 0.29917280
Iteration 12, loss = 0.29906183
Iteration 13, loss = 0.29898394
Training loss did not improve more than tol=0.010000 for 10 consecutive epochs. Stopping.


MLPClassifier(hidden_layer_sizes=(53, 53), max_iter=500, random_state=42,
              solver='sgd', tol=0.01, verbose=True)

In [19]:
predict = mlp_classifier.predict(x_test)
print(f'Função de Ativação na Camada de Saída: {mlp_classifier.out_activation_}')
acuracia = accuracy_score(y_test, predict)
matriz_de_confusao = confusion_matrix(y_test, predict)

print(f'Acurácia: {acuracia}')
print(matriz_de_confusao)

Função de Ativação na Camada de Saída: logistic
Acurácia: 0.8692274386095132
[[ 5094  7561]
 [ 3010 65170]]


# <font color='red' style='font-size: 40px;'> 4.0) Redes Neurais Artificiais - Regressão  </font>



## **Força Elástica**

In [26]:
import pandas as pd 
import numpy as np
import scipy as sp 
import matplotlib.pyplot as plt
import seaborn as sns 
import warnings
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler
from sklearn.model_selection import train_test_split, StratifiedKFold, cross_val_score
from sklearn.compose import ColumnTransformer
from sklearn.neural_network import MLPRegressor
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

%matplotlib inline
sns.set()
plt.rcParams['font.family'] = 'Arial'
plt.rcParams['font.size'] = '14'
plt.rcParams['figure.figsize'] = [10, 5]
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 100)
pd.set_option('display.float_format', lambda x: '%.2f' % x) # Tira os números do formato de Notação Científica
np.set_printoptions(suppress=True) # Tira os números do formato de Notação Científica em Numpy Arrays
warnings.filterwarnings('ignore') # Remove Warnings
warnings.simplefilter(action='ignore', category=FutureWarning) # Retira Future Warnings

In [24]:
rigidez = pd.read_csv('mola.csv', sep=',', encoding='iso-8859-1')
rigidez.drop('cargas', axis=1, inplace = True)
rigidez.drop('Lo',axis=1, inplace = True)
print(rigidez.shape)
rigidez.head()

(26, 4)


forca     L    x     K
0   0.10 29.90 0.00 34.48
1   0.15 31.30 0.00 34.88
2   0.23 34.00 0.01 32.86
3   0.45 41.40 0.01 31.25
4   0.55 44.20 0.02 31.98

In [25]:
x = rigidez.iloc[:, 2:3].values
y = rigidez.iloc[:, 0:1].values


In [28]:
mlp_regressor = MLPRegressor(hidden_layer_sizes = (26, 26), activation = 'identity', solver = 'lbfgs', alpha = 0.001, batch_size= 'auto', 
learning_rate='constant', learning_rate_init=0.001, max_iter = 500, tol = 0.01, random_state = 42, verbose = True)
mlp_regressor.fit(x, y) 

MLPRegressor(activation='identity', alpha=0.001, hidden_layer_sizes=(26, 26),
             max_iter=500, random_state=42, solver='lbfgs', tol=0.01,
             verbose=True)

In [30]:
print(f'Número de Camadas (contando camadas de entrada, camadas ocultas e camadas de saída): {mlp_regressor.n_layers_}')

Número de Camadas (contando camadas de entrada, camadas ocultas e camadas de saída): 4
